merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ThreadedSimulationExtentReporter.java

```diff
@@ -0,0 +1,83 @@
+package gov.nasa.jpl.aerie.merlin.server.services;
+
+import gov.nasa.jpl.aerie.merlin.driver.SimulationDriver;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ThreadedSimulationExtentReporter implements SimulationDriver.SimulationExtentReporter, AutoCloseable {
+  private final AtomicReference<Duration> latestExtent;
+  private final AutoCloseable worker;
+
+  private ThreadedSimulationExtentReporter(final AtomicReference<Duration> latestExtent, final AutoCloseable worker) {
+    this.latestExtent = latestExtent;
+    this.worker = worker;
+  }
+
+  public static ThreadedSimulationExtentReporter spawn(
+      final String threadName,
+      final SimulationDriver.SimulationExtentReporter listener,
+      final long minimumPollPeriodSeconds
+  ) {
+    final var latestExtent = new AtomicReference<>(Duration.ZERO);
+    final var thread = new Thread(new ThreadedSimulationExtentReporter.Worker(latestExtent, listener, minimumPollPeriodSeconds));
+    thread.setName(threadName);
+    thread.start();
+    return new ThreadedSimulationExtentReporter(latestExtent, thread::interrupt);
+  }
+
+  @Override
+  public void report(final Duration extent) {
+    this.latestExtent.set(extent);
+  }
+
+  @Override
+  public void close() {
+    try {
+      this.worker.close();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static class Worker implements Runnable {
+    private final AtomicReference<Duration> latestExtent;
+    private final SimulationDriver.SimulationExtentReporter listener;
+    private final long minimumPollPeriodMillis;
+
+    Worker(
+        final AtomicReference<Duration> latestExtent,
+        final SimulationDriver.SimulationExtentReporter listener,
+        final long minimumPollPeriodMillis
+    ) {
+      this.latestExtent = latestExtent;
+      this.listener = listener;
+      this.minimumPollPeriodMillis = minimumPollPeriodMillis;
+    }
+
+    @Override
+    public void run() {
+      try {
+        listener.report(latestExtent.get());
+      } catch (Exception e) {
+        e.printStackTrace(System.err);
+      }
+
+      var lastPollTime = System.nanoTime();
+      while (!Thread.currentThread().isInterrupted()) {
+        try {
+          final var millisSinceLastPoll = (System.nanoTime() - lastPollTime) / 1_000_000;
+          if (millisSinceLastPoll < minimumPollPeriodMillis) {
+            Thread.sleep(minimumPollPeriodMillis - millisSinceLastPoll);
+          }
```

Comment:
Could you elaborate on what you mean by a spurious wakeup?

Reply:
Sure! A [spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup) is when a thread is woken up but it's condition is not satisfied. This is a concern that is explicitly called out in [the docs for `Object.wait()`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Object.html#wait(long,int)). With `Thread.sleep`, this can happen if the OS decides to schedule the sleeping thread early.

Reply:
I see, thanks

> But spurious wakeup don't happen for no reason: they usually happen because, in between the time when the condition variable was signaled and when the waiting thread finally ran, another thread ran and changed the condition

Since the condition we're waiting for is "time progressing", it does not appear to me that we are vulnerable to spurious wakeups. There is no way for our condition to be triggered, and then negated by another thread.

https://stackoverflow.com/a/2389016

That said, I'm not opposed to making it a while loop - I'll give it a shot